utils/debug-grammar.js

```diff
@@ -0,0 +1,62 @@
+/**
+ * Add env variable `VSCODE_TEXTMATE_DEBUG=true` to run script with extra debugging info
+ * Example: `VSCODE_TEXTMATE_DEBUG=true node utils/debug-grammar.js example.apib`
+ */
+
+const fs = require('fs');
+const path = require('path');
+const vsctm = require('vscode-textmate');
+const oniguruma = require('vscode-oniguruma');
+
+const FILE_PATH = process.argv[2];
+
+const fileContents = fs.readFileSync(FILE_PATH, { encoding: 'utf-8' });
+const fileLines = fileContents.split(/\r\n|\r|\n/);
+
+const scopeToGrammar = {
+  'text.html.markdown.source.gfm.apib': getPath('../syntax/APIBlueprint.tmLanguage'),
+  'text.html.markdown.source.gfm.mson': getPath('../syntax/MSON.tmLanguage'),
+  'source.js': getPath('../syntax/JavaScript.tmLanguage'),
+  'text.html.basic': getPath('../syntax/html.tmLanguage'),
+};
+
+const wasmBin = fs.readFileSync(getPath('../node_modules/vscode-oniguruma/release/onig.wasm')).buffer;
+const vscodeOnigurumaLib = oniguruma.loadWASM(wasmBin).then(() => ({
+  createOnigScanner(patterns) { return new oniguruma.OnigScanner(patterns); },
+  createOnigString(s) { return new oniguruma.OnigString(s); },
+}));
+
+// Create a registry that can create a grammar from a scope name.
+const registry = new vsctm.Registry({
+  onigLib: vscodeOnigurumaLib,
+  loadGrammar: (scopeName) => {
+    const pathToGrammar = scopeToGrammar[scopeName];
+
+    if (pathToGrammar) {
+      return fs.promises.readFile(pathToGrammar, { encoding: 'utf-8' }).then(data => vsctm.parseRawGrammar(data, pathToGrammar));
+    }
+    console.log(`Unknown scope name: ${scopeName}`);
+    return null;
+  },
+});
+
+// Load the JavaScript grammar and any other grammars included by it async.
+registry.loadGrammar('text.html.markdown.source.gfm.apib').then(grammar => {
```

Comment:
This script is intended to help a developer with debugging of apib grammar in `.tmLanguage` files. It mostly repeats the example from https://github.com/microsoft/vscode-textmate and it helped me a lot with the solution, so I decided to keep it for the future.

**Basic usage**: `node utils/debug-grammar.js example.apib`, assuming that you have an APIB doc snippet in `example.apib`. I do not recommend to put a large doc in this file since it will bloat your stdout. Just focus on some small pieces of what you need.

Imagine the next example:

```markdown
# My API

# Data Structures

## User

+ name (string)

## Users [POST /users]

+ Response 200
```

Script output will be:
```
Tokenizing line: # My API
 - token from 0 to 1 (#) with scopes text.html.markdown.source.gfm.apib, markup.heading.heading-1.gfm, markup.heading.marker.gfm
 - token from 1 to 2 ( ) with scopes text.html.markdown.source.gfm.apib, markup.heading.heading-1.gfm, markup.heading.space.gfm
 - token from 2 to 8 (My API) with scopes text.html.markdown.source.gfm.apib, markup.heading.heading-1.gfm

Tokenizing line: 
 - token from 0 to 1 () with scopes text.html.markdown.source.gfm.apib

Tokenizing line: # Data Structures
 - token from 0 to 17 (# Data Structures) with scopes text.html.markdown.source.gfm.apib, markup.heading.markdown, punctuation.definition.heading.markdown

Tokenizing line: 
 - token from 0 to 1 () with scopes text.html.markdown.source.gfm.apib

Tokenizing line: ## User
 - token from 0 to 7 (## User) with scopes text.html.markdown.source.gfm.apib, markup.heading.markdown, punctuation.definition.heading.markdown

Tokenizing line: 
 - token from 0 to 1 () with scopes text.html.markdown.source.gfm.apib

Tokenizing line: + name (string)
 - token from 0 to 1 (+) with scopes text.html.markdown.source.gfm.apib, mson.field.name, variable.unordered.list.gfm
 - token from 1 to 6 ( name) with scopes text.html.markdown.source.gfm.apib, mson.field.name
 - token from 6 to 7 ( ) with scopes text.html.markdown.source.gfm.apib
 - token from 7 to 8 (() with scopes text.html.markdown.source.gfm.apib, mson.field.attributes
 - token from 8 to 14 (string) with scopes text.html.markdown.source.gfm.apib, mson.field.attributes, mson-element-type, entity.name.class
 - token from 14 to 15 ()) with scopes text.html.markdown.source.gfm.apib, mson.field.attributes

Tokenizing line: 
 - token from 0 to 1 () with scopes text.html.markdown.source.gfm.apib

Tokenizing line: ## Users [POST /users]
 - token from 0 to 10 (## Users [) with scopes text.html.markdown.source.gfm.apib, markup.heading.markdown, punctuation.definition.heading.markdown
 - token from 10 to 14 (POST) with scopes text.html.markdown.source.gfm.apib, markup.heading.markdown, punctuation.definition.heading.markdown, markup.quote
 - token from 14 to 21 ( /users) with scopes text.html.markdown.source.gfm.apib, markup.heading.markdown, punctuation.definition.heading.markdown, meta.link.inet.markdown, markup.underline.link.markdown
 - token from 21 to 23 (]) with scopes text.html.markdown.source.gfm.apib, markup.heading.markdown, punctuation.definition.heading.markdown

Tokenizing line: 
 - token from 0 to 1 () with scopes text.html.markdown.source.gfm.apib

Tokenizing line: + Response 200
 - token from 0 to 1 (+) with scopes text.html.markdown.source.gfm.apib, blueprint.response, variable.unordered.list.gfm
 - token from 1 to 2 ( ) with scopes text.html.markdown.source.gfm.apib, blueprint.response
 - token from 2 to 10 (Response) with scopes text.html.markdown.source.gfm.apib, blueprint.response, constant.numeric
 - token from 10 to 14 ( 200) with scopes text.html.markdown.source.gfm.apib, blueprint.response, constant.numeric

Tokenizing line: 
 - token from 0 to 1 () with scopes text.html.markdown.source.gfm.apib, blueprint.response
```

As you can see, you have pretty clear mapping between a line (or even part of the line) and a corresponding token.
If some part of a doc is not highlighted as you want it to be, most likely it will have scope `text.html.markdown.source.gfm.apib` (aka no semantics).

**Extended usage:**
Basic output of assigned tokens was not enough for me, but, thanks God, I spotted a debug mode in the `vscode-textmate` package.
To activate it provide env variable `VSCODE_TEXTMATE_DEBUG=true`.

If activated, it outputs for a specific line a set of regular expressions it MAY match, the resulting selected regular expression, and some additional info

Part of the extended output:

```
@@scanNext 0: |# Data Structures\n|
  scanning for (linePos: 0, anchorPosition: -1)
   - 2: ^#{1,6}\s+[Gg]roup.*\n?
   - 3: ^(#{1,6}\s+[Dd]ata [Ss]tructures)
   - 973: ^(#{1,6}\s+[Ii]mport\s+(.+))$
   - 978: ^#{1,6}\s+.*\[(HEAD|GET|PUT|POST|PATCH|DELETE)(.*)\]\n?
   - 981: ^#{1,6}\s+.*\[(.*)\]\n?
   - 983: ^#{1,6}\s+(HEAD|GET|PUT|POST|PATCH|DELETE)(.*)\n?
   - 985: ^([-+*]) ([Rr]equest)(\s+([^\]\[\(\)]+))?(\s*\(([^\)]*)json\))$
   - 1007: ^([-+*]) ([Rr]equest)(\s+([^\]\[\(\)]+))?(\s*\(([^\)]*)xml\))$
   - 1014: ^([-+*]) ([Rr]equest)(\s+([^\]\[\(\)]+))?(\s*\(([^\)]*)\))?$
   - 1019: ^([-+*]) ([Rr]esponse)(\s+(\d+))?(\s*\(([^\)]*)json\))$
   - 1026: ^([-+*]) ([Rr]esponse)(\s+(\d+))?(\s*\(([^\)]*)xml\))$
   - 1033: ^([-+*]) ([Rr]esponse)(\s+(\d+))?(\s*\(([^\)]*)\))?$
   - 996: ^(\s*)([-+*]) ([Pp]arameters)
   - 999: ^(\s+)([-+*]) ([Aa]ttributes)(\s*)(\(.+\))?$
   ........... // here is more, I omitted it
matched rule id: 3 from 0 to 17
  pushing t#3 @ APIBlueprint.tmLanguage:469 - ^(#{1,6}\s+[Dd]ata [Ss]tructures)
  token: |# Data Structures|
      * text.html.markdown.source.gfm.apib
      * markup.heading.markdown
      * punctuation.definition.heading.markdown
```

This extended mode helped me to better understand (not fully, just better) how the stuff works. I believe it can help us in other possible problems or improvements.

Reply:
Gorgeous.